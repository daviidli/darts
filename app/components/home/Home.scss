.home {
	.title {
		font-weight: 300;
		text-align: center;
		padding: 1.5em;
	}

	.button {
		width: 12em;

		&:hover {
			background-color: rgba(255, 255, 255, 0.3);
		}
	}
}
