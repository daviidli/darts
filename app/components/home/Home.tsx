import React, { useEffect } from 'react';
import { useHistory } from 'react-router-dom';
import { Grid, Typography, Button } from '@material-ui/core';
import SerialPort from 'serialport';
import routes from '../../constants/routes.json';
import styles from './Home.scss';
import Options, { optionsItems } from '../options/Options';

export type Props = {
	serialPort: string;
	setSerialPort: (port: string) => {};
};

const Home = (props: Props) => {
	const { serialPort, setSerialPort } = props;
	const history = useHistory();

	useEffect(() => {
		const fetchPorts = async () => {
			try {
				let p = await SerialPort.list();
				p = p.filter(port => port.productId);
				if (p.length) {
					setSerialPort(p[0].path);
				}
			} catch (e) {
				// continue regardless of error
				// todo: show error notification in future
			}
		};

		if (serialPort === '') {
			fetchPorts();
		}
	}, [serialPort, setSerialPort]);

	const gameModes: optionsItems[] = [
		{
			label: "01's",
			onClick: () => history.push(routes.OONE)
		}
	];

	return (
		<Grid
			container
			direction="column"
			justify="center"
			alignItems="center"
			spacing={3}
			className={styles.home}
		>
			<Grid item>
				<Typography className={styles.title} variant="h2">
					Darts
				</Typography>
			</Grid>
			<Grid item>
				<Options items={gameModes} />
			</Grid>
			<Grid item>
				<Button
					className={styles.button}
					variant="contained"
					color="secondary"
					disableElevation
					onClick={() => history.push(routes.SETTINGS)}
				>
					Settings
				</Button>
			</Grid>
		</Grid>
	);
};

export default Home;
