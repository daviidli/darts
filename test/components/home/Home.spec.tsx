import { spy, stub } from 'sinon';
import React from 'react';
import Enzyme, { shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import SerialPort from 'serialport';
import renderer from 'react-test-renderer';
import Home, { Props } from '../../../app/components/home/Home';

Enzyme.configure({ adapter: new Adapter() });

const mockPush = spy();
jest.mock('react-router-dom', () => ({
	...jest.requireActual('react-router-dom'),
	useHistory: () => ({
		push: mockPush
	})
}));

const initialProps: Props = {
	serialPort: 'port1',
	setSerialPort: () => ({})
};

const setup = (props: Props = initialProps) => {
	const wrapper = shallow(<Home {...props} />);
	return {
		wrapper,
		title: wrapper.find('.title'),
		button: wrapper.find('.button')
	};
};

const setupMount = (props: Props = initialProps) => {
	const wrapper = mount(<Home {...props} />);
	return {
		wrapper,
		title: wrapper.find('.title'),
		button: wrapper.find('.button')
	};
};

describe('Home component', () => {
	it('should match the snapshot', () => {
		const tree = renderer.create(<Home {...initialProps} />).toJSON();
		expect(tree).toMatchSnapshot();
	});

	it('should go to settings screen on click', () => {
		const { button } = setup();
		expect(button.at(0).text()).toMatch('Settings');
		button.at(0).simulate('click');
		expect(mockPush.callCount).toBe(1);
		expect(mockPush.calledWith('/settings')).toBeTruthy();
	});

	it('should not select any serial ports due to SerialPort.list() return of []', () => {
		const setSerialPort = spy();
		const props: Props = {
			serialPort: '',
			setSerialPort
		};

		setupMount(props);

		return new Promise(resolve => {
			setImmediate(() => {
				expect(setSerialPort.notCalled).toBeTruthy();
				resolve();
			});
		});
	});

	it('should select the first available serial port', () => {
		const path = 'port123';
		stub(SerialPort, 'list').returns(
			Promise.resolve([{ path, productId: '123' }])
		);
		const setSerialPort = spy();
		const props: Props = {
			serialPort: '',
			setSerialPort
		};

		setupMount(props);

		return new Promise(resolve => {
			setImmediate(() => {
				expect(setSerialPort.callCount).toBe(1);
				expect(setSerialPort.calledWith(path)).toBeTruthy();
				resolve();
			});
		});
	});

	it('should not select any ports as one is already selected', () => {
		const setSerialPortMock = spy();
		const props = { ...initialProps };
		props.setSerialPort = setSerialPortMock;

		setupMount(props);

		return new Promise(resolve => {
			setImmediate(() => {
				expect(setSerialPortMock.callCount).toBe(0);
				resolve();
			});
		});
	});
});
