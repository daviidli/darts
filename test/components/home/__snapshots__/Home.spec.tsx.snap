// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home component should match the snapshot 1`] = `
<div
  className="MuiGrid-root home MuiGrid-container MuiGrid-spacing-xs-3 MuiGrid-direction-xs-column MuiGrid-align-items-xs-center MuiGrid-justify-xs-center"
>
  <div
    className="MuiGrid-root MuiGrid-item"
  >
    <h2
      className="MuiTypography-root title MuiTypography-h2"
    >
      Darts
    </h2>
  </div>
  <div
    className="MuiGrid-root MuiGrid-item"
  >
    <div
      className="MuiGrid-root options MuiGrid-container MuiGrid-justify-xs-center"
    >
      <div
        className="MuiGrid-root MuiGrid-item"
      >
        <button
          className="MuiButtonBase-root MuiButton-root MuiButton-contained button MuiButton-containedPrimary"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <span
            className="MuiButton-label"
          >
            01's
          </span>
        </button>
      </div>
    </div>
  </div>
  <div
    className="MuiGrid-root MuiGrid-item"
  >
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained button MuiButton-containedSecondary MuiButton-disableElevation"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-label"
      >
        Settings
      </span>
    </button>
  </div>
</div>
`;
